README.md
# Slider Carousel

A new slider carousel widget with custom scaling and item spacing

## Example

    SliderCarousel(
      itemCount: 5,
      itemBuilder: (context, idx) {
        return Container(
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            color: Colors.blue,
          ),
        );
      },
      offsetCenter: 0.75,
      scaleCenter: 1.25,
      itemWidth: 150,
      itemHeight: 150,
    );


![](https://github.com/devdanpre/flutter_slidercarousel/blob/main/example/images/screenshot.gif)





CHANGELOG.md
# v0.0.2 - Add Null safety (@hsnbsst)

add Null safety

# v0.0.1 - February 2021

Initial version.

test/flutter_slidercarousel_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_slidercarousel/flutter_slidercarousel.dart';

void main() {
  testWidgets('Default SliderCarousel', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MaterialApp(
        home: SliderCarousel(
            itemBuilder: (context, index) {
              return Text("0");
            },
            itemCount: 10)));

    expect(find.text("0", skipOffstage: false), findsOneWidget);
  });



}


lib/src/custom_layout.dart
part of 'slidercarousel.dart';

abstract class _CustomLayoutStateBase<T extends _SubSliderCarousel>
    extends State<T> with SingleTickerProviderStateMixin {
  double _sliderWidth = 0;
  double _sliderHeight = 0;
  late Animation<double> _animation;
  late AnimationController _animationController;
  int _startIndex = 0;
  int _animationCount = 5;

  @override
  void initState() {
    if (widget.itemWidth == null) {
      throw new Exception(
          "==============\n\nwidget.itemWith must not be null when use slider.\n========\n");
    }

    _createAnimationController();
    widget.controller!.addListener(_onController);
    super.initState();
  }

  void _createAnimationController() {
    _animationController = new AnimationController(vsync: this, value: 0.5);
    Tween<double> tween = new Tween(begin: 0.0, end: 1.0);
    _animation = tween.animate(_animationController);
  }

  @override
  void didChangeDependencies() {
    WidgetsBinding.instance!.addPostFrameCallback(_getSize);
    super.didChangeDependencies();
  }

  void _getSize(_) {
    afterRender();
  }

  @mustCallSuper
  void afterRender() {
    RenderObject? renderObject = context.findRenderObject();
    Size size;
    if (renderObject == null) {
      size = Size(0, 0);
    } else {
      size = renderObject.paintBounds.size;
    }

    _sliderWidth = size.width;
    _sliderHeight = size.height;
    if (mounted) {
      setState(() {});
    }
  }

  @override
  void didUpdateWidget(T oldWidget) {
    if (widget.controller != oldWidget.controller) {
      oldWidget.controller!.removeListener(_onController);
      widget.controller!.addListener(_onController);
    }

    if (widget.loop != oldWidget.loop) {
      if (!widget.loop!) {
        _currentIndex = _ensureIndex(_currentIndex);
      }
    }

    super.didUpdateWidget(oldWidget);
  }

  int _ensureIndex(int index) {
    index = index % widget.itemCount!;
    if (index < 0) {
      index += widget.itemCount!;
    }
    return index;
  }

  @override
  void dispose() {
    widget.controller!.removeListener(_onController);
    _animationController.dispose();
    super.dispose();
  }

  Widget _buildItem(int i, int realIndex, double animationValue);

  Widget _buildContainer(List<Widget> list) {
    return new Stack(
      children: list,
    );
  }

  Widget _buildAnimation(BuildContext context, Widget? w) {
    List<Widget> list = [];

    double animationValue = _animation.value;

    for (int i = 0; i < _animationCount; ++i) {
      int realIndex = _currentIndex + i + _startIndex;
      realIndex = realIndex % widget.itemCount!;
      if (realIndex < 0) {
        realIndex += widget.itemCount!;
      }

      list.add(_buildItem(i, realIndex, animationValue));
    }

    return new GestureDetector(
      behavior: HitTestBehavior.opaque,
      onPanStart: _onPanStart,
      onPanEnd: _onPanEnd,
      onPanUpdate: _onPanUpdate,
      child: new ClipRect(
        child: new Center(
          child: _buildContainer(list),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    if (_animationCount == null) {
      return new Container();
    }
    return new AnimatedBuilder(
      animation: _animationController,
      builder: _buildAnimation,
    );
  }

  double? _currentValue;
  double? _currentPos;

  bool _lockScroll = false;

  void _move(double position, {int? nextIndex}) async {
    if (_lockScroll) return;
    try {
      _lockScroll = true;
      await _animationController.animateTo(position,
          duration: new Duration(milliseconds: widget.duration!),
          curve: widget.curve!);
      if (nextIndex != null) {
        widget.onIndexChanged!(widget.getCorrectIndex(nextIndex));
      }
    } catch (e) {
      print(e);
    } finally {
      if (nextIndex != null) {
        try {
          _animationController.value = 0.5;
        } catch (e) {
          print(e);
        }

        _currentIndex = nextIndex;
      }
      _lockScroll = false;
    }
  }

  int _nextIndex() {
    int index = _currentIndex + 1;
    if (!widget.loop! && index >= widget.itemCount! - 1) {
      return widget.itemCount! - 1;
    }
    return index;
  }

  int _prevIndex() {
    int index = _currentIndex - 1;
    if (!widget.loop! && index < 0) {
      return 0;
    }
    return index;
  }

  void _onController() {
    switch (widget.controller!.event) {
      case IndexController.PREVIOUS:
        int prevIndex = _prevIndex();
        if (prevIndex == _currentIndex) return;
        _move(1.0, nextIndex: prevIndex);
        break;
      case IndexController.NEXT:
        int nextIndex = _nextIndex();
        if (nextIndex == _currentIndex) return;
        _move(0.0, nextIndex: nextIndex);
        break;
      case IndexController.MOVE:
        throw new Exception(
            "Custom layout does not support SliderCarouselControllerEvent.MOVE_INDEX yet!");
      case SliderCarouselController.STOP_AUTOPLAY:
      case SliderCarouselController.START_AUTOPLAY:
        break;
    }
  }

  void _onPanEnd(DragEndDetails details) {
    if (_lockScroll) return;

    double velocity = widget.scrollDirection == Axis.horizontal
        ? details.velocity.pixelsPerSecond.dx
        : details.velocity.pixelsPerSecond.dy;

    if (_animationController.value >= 0.75 || velocity > 500.0) {
      if (_currentIndex <= 0 && !widget.loop!) {
        return;
      }
      _move(1.0, nextIndex: _currentIndex - 1);
    } else if (_animationController.value < 0.25 || velocity < -500.0) {
      if (_currentIndex >= widget.itemCount! - 1 && !widget.loop!) {
        return;
      }
      _move(0.0, nextIndex: _currentIndex + 1);
    } else {
      _move(0.5);
    }
  }

  void _onPanStart(DragStartDetails details) {
    if (_lockScroll) return;
    _currentValue = _animationController.value;
    _currentPos = widget.scrollDirection == Axis.horizontal
        ? details.globalPosition.dx
        : details.globalPosition.dy;
  }

  void _onPanUpdate(DragUpdateDetails details) {
    if (_lockScroll) return;
    double value = _currentValue! +
        ((widget.scrollDirection == Axis.horizontal
                    ? details.globalPosition.dx
                    : details.globalPosition.dy) -
                _currentPos!) /
            _sliderWidth /
            2;
    // no loop ?
    if (!widget.loop!) {
      if (_currentIndex >= widget.itemCount! - 1) {
        if (value < 0.5) {
          value = 0.5;
        }
      } else if (_currentIndex <= 0) {
        if (value > 0.5) {
          value = 0.5;
        }
      }
    }

    _animationController.value = value;
  }

  int _currentIndex = 0;
}

double _getValue(List<double> values, double animationValue, int index) {
  double s = values[index];
  if (animationValue >= 0.5) {
    if (index < values.length - 1) {
      s = s + (values[index + 1] - s) * (animationValue - 0.5) * 2.0;
    }
  } else {
    if (index != 0) {
      s = s - (s - values[index - 1]) * (0.5 - animationValue) * 2.0;
    }
  }
  return s;
}

Offset _getOffsetValue(List<Offset> values, double animationValue, int index) {
  Offset s = values[index];
  double dx = s.dx;
  double dy = s.dy;
  if (animationValue >= 0.5) {
    if (index < values.length - 1) {
      dx = dx + (values[index + 1].dx - dx) * (animationValue - 0.5) * 2.0;
      dy = dy + (values[index + 1].dy - dy) * (animationValue - 0.5) * 2.0;
    }
  } else {
    if (index != 0) {
      dx = dx - (dx - values[index - 1].dx) * (0.5 - animationValue) * 2.0;
      dy = dy - (dy - values[index - 1].dy) * (0.5 - animationValue) * 2.0;
    }
  }
  return new Offset(dx, dy);
}

abstract class TransformBuilder<T> {
  List<T> values;
  TransformBuilder({required this.values});
  Widget build(int i, double animationValue, Widget widget);
}

class ScaleTransformBuilder extends TransformBuilder<double> {
  final Alignment alignment;
  ScaleTransformBuilder(
      {required List<double> values, this.alignment: Alignment.center})
      : super(values: values);

  Widget build(int i, double animationValue, Widget widget) {
    double s = _getValue(values, animationValue, i);
    return new Transform.scale(scale: s, child: widget);
  }
}

class OpacityTransformBuilder extends TransformBuilder<double> {
  OpacityTransformBuilder({required List<double> values})
      : super(values: values);

  Widget build(int i, double animationValue, Widget widget) {
    double v = _getValue(values, animationValue, i);
    return new Opacity(
      opacity: v,
      child: widget,
    );
  }
}

class RotateTransformBuilder extends TransformBuilder<double> {
  RotateTransformBuilder({required List<double> values})
      : super(values: values);

  Widget build(int i, double animationValue, Widget widget) {
    double v = _getValue(values, animationValue, i);
    return new Transform.rotate(
      angle: v,
      child: widget,
    );
  }
}

class TranslateTransformBuilder extends TransformBuilder<Offset> {
  TranslateTransformBuilder({required List<Offset> values})
      : super(values: values);

  @override
  Widget build(int i, double animationValue, Widget widget) {
    Offset s = _getOffsetValue(values, animationValue, i);
    return new Transform.translate(
      offset: s,
      child: widget,
    );
  }
}

class CustomLayoutOption {
  final List<TransformBuilder> builders = [];
  final int startIndex;
  final int stateCount;

  CustomLayoutOption({required this.stateCount, required this.startIndex});

  CustomLayoutOption addOpacity(List<double> values) {
    builders.add(new OpacityTransformBuilder(values: values));
    return this;
  }

  CustomLayoutOption addTranslate(List<Offset> values) {
    builders.add(new TranslateTransformBuilder(values: values));
    return this;
  }

  CustomLayoutOption addScale(List<double> values, Alignment alignment) {
    builders
        .add(new ScaleTransformBuilder(values: values, alignment: alignment));
    return this;
  }

  CustomLayoutOption addRotate(List<double> values) {
    builders.add(new RotateTransformBuilder(values: values));
    return this;
  }
}

class _CustomLayoutSliderCarousel extends _SubSliderCarousel {
  final CustomLayoutOption option;

  _CustomLayoutSliderCarousel(
      {required this.option,
      double? itemWidth,
      bool? loop,
      double? itemHeight,
      ValueChanged<int>? onIndexChanged,
      Key? key,
      IndexedWidgetBuilder? itemBuilder,
      Curve? curve,
      int? duration,
      int? index,
      int? itemCount,
      Axis? scrollDirection,
      SliderCarouselController? controller})
      : super(
            loop: loop!,
            onIndexChanged: onIndexChanged!,
            itemWidth: itemWidth!,
            itemHeight: itemHeight!,
            key: key!,
            itemBuilder: itemBuilder!,
            curve: curve!,
            duration: duration!,
            index: index!,
            itemCount: itemCount!,
            controller: controller!,
            scrollDirection: scrollDirection!);

  @override
  State<StatefulWidget> createState() {
    return new _CustomLayoutState();
  }
}

class _CustomLayoutState
    extends _CustomLayoutStateBase<_CustomLayoutSliderCarousel> {
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _startIndex = widget.option.startIndex;
    _animationCount = widget.option.stateCount;
  }

  @override
  void didUpdateWidget(_CustomLayoutSliderCarousel oldWidget) {
    _startIndex = widget.option.startIndex;
    _animationCount = widget.option.stateCount;
    super.didUpdateWidget(oldWidget);
  }

  @override
  Widget _buildItem(int index, int realIndex, double animationValue) {
    List<TransformBuilder> builders = widget.option.builders;

    Widget child = new SizedBox(
        width: widget.itemWidth ?? double.infinity,
        height: widget.itemHeight ?? double.infinity,
        child: widget.itemBuilder!(context, realIndex));

    for (int i = builders.length - 1; i >= 0; --i) {
      TransformBuilder builder = builders[i];
      child = builder.build(index, animationValue, child);
    }

    return child;
  }
}


lib/src/slidercarousel.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

import '../../transformer_page_view/index_controller.dart';
import '../../transformer_page_view/transformer_page_view.dart';
import 'slider_controller.dart';

part 'custom_layout.dart';

typedef void SliderCarouselOnTap(int index);

typedef Widget SliderCarouselDataBuilder(
    BuildContext context, dynamic data, int index);

/// default auto play
const int defaultAutoplayDelayMs = 3000;

///  Default auto play transition duration (in millisecond)
const int defaultAutoplayTransactionDuration = 300;

class SliderCarousel extends StatefulWidget {
  final double? itemHeight;
  final double? itemWidth;
  final double offsetCenter;
  final double scaleCenter;
  final IndexedWidgetBuilder? itemBuilder;
  final int itemCount;
  final ValueChanged<int>? onIndexChanged;
  final bool autoplay;
  final int autoplayDelay;
  final bool autoplayDisableOnInteraction;
  final int duration;
  final Axis scrollDirection;
  final Curve curve;
  final bool loop;
  final int? index;
  final SliderCarouselController? controller;

  SliderCarousel({
    Key? key,
    this.curve: Curves.ease,
    this.duration: defaultAutoplayTransactionDuration,
    this.controller,
    this.onIndexChanged,
    this.itemHeight,
    this.itemWidth,
    this.itemBuilder,
    this.index,
    this.loop: true,
    this.autoplay: false,
    this.autoplayDelay: defaultAutoplayDelayMs,
    this.autoplayDisableOnInteraction: true,
    required this.itemCount,
    this.scrollDirection: Axis.horizontal,
    this.offsetCenter = 0.75,
    this.scaleCenter = 1.25,
  }) : super(key: key);

  factory SliderCarousel.children({
    List<Widget>? children,
    bool autoplay: false,
    int autoplayDelay: defaultAutoplayDelayMs,
    bool autoplayDisableOnInteraction: true,
    int duration: defaultAutoplayTransactionDuration,
    ValueChanged<int>? onIndexChanged,
    int? index,
    bool loop: true,
    Curve curve: Curves.ease,
    Axis scrollDirection: Axis.horizontal,
    SliderCarouselController? controller,
    Key? key,
    double? itemHeight,
    double? itemWidth,
    double? offsetCenter,
    double? scaleCenter,
  }) {
    assert(children != null, "children must not be null");

    return new SliderCarousel(
        itemHeight: itemHeight,
        itemWidth: itemWidth,
        autoplay: autoplay,
        autoplayDelay: autoplayDelay,
        autoplayDisableOnInteraction: autoplayDisableOnInteraction,
        duration: duration,
        onIndexChanged: onIndexChanged,
        index: index,
        curve: curve,
        scrollDirection: scrollDirection,
        controller: controller,
        loop: loop,
        key: key,
        itemBuilder: (BuildContext context, int index) {
          return children![index];
        },
        offsetCenter: offsetCenter!,
        scaleCenter: scaleCenter!,
        itemCount: children!.length);
  }

  factory SliderCarousel.list({
    List? list,
    required CustomLayoutOption customLayoutOption,
    SliderCarouselDataBuilder? builder,
    bool autoplay: false,
    int autoplayDelay: defaultAutoplayDelayMs,
    bool reverse: false,
    bool autoplayDisableOnInteraction: true,
    int duration: defaultAutoplayTransactionDuration,
    ValueChanged<int>? onIndexChanged,
    int? index,
    bool loop: true,
    Curve curve: Curves.ease,
    Axis scrollDirection: Axis.horizontal,
    SliderCarouselController? controller,
    Key? key,
    double? itemHeight,
    double? itemWidth,
    double? offsetCenter,
    double? scaleCenter,
  }) {
    return new SliderCarousel(
        itemHeight: itemHeight,
        itemWidth: itemWidth,
        autoplay: autoplay,
        autoplayDelay: autoplayDelay,
        autoplayDisableOnInteraction: autoplayDisableOnInteraction,
        duration: duration,
        onIndexChanged: onIndexChanged,
        index: index,
        curve: curve,
        key: key,
        scrollDirection: scrollDirection,
        controller: controller,
        loop: loop,
        itemBuilder: (BuildContext context, int index) {
          return builder!(context, list![index], index);
        },
        offsetCenter: offsetCenter!,
        scaleCenter: scaleCenter!,
        itemCount: list!.length);
  }

  @override
  State<StatefulWidget> createState() {
    return new _SliderCarouselState();
  }
}

abstract class _SliderCarouselTimerMixin extends State<SliderCarousel> {
  Timer? _timer;

  late SliderCarouselController _controller;

  @override
  void initState() {
    _controller = widget.controller!;
    // ignore: unnecessary_null_comparison
    if (_controller == null) {
      _controller = new SliderCarouselController();
    }
    _controller.addListener(_onController);
    _handleAutoplay();
    super.initState();
  }

  void _onController() {
    switch (_controller.event) {
      case SliderCarouselController.START_AUTOPLAY:
        {
          if (_timer == null) {
            _startAutoplay();
          }
        }
        break;
      case SliderCarouselController.STOP_AUTOPLAY:
        {
          if (_timer != null) {
            _stopAutoplay();
          }
        }
        break;
    }
  }

  @override
  void didUpdateWidget(SliderCarousel oldWidget) {
    if (_controller != oldWidget.controller) {
      if (oldWidget.controller != null) {
        oldWidget.controller!.removeListener(_onController);
        _controller = oldWidget.controller!;
        _controller.addListener(_onController);
      }
    }
    _handleAutoplay();
    super.didUpdateWidget(oldWidget);
  }

  @override
  void dispose() {
    _controller.removeListener(_onController);

    _stopAutoplay();
    super.dispose();
  }

  bool _autoplayEnabled() {
    return _controller.autoplay ?? widget.autoplay;
  }

  void _handleAutoplay() {
    if (_autoplayEnabled() && _timer != null) return;
    _stopAutoplay();
    if (_autoplayEnabled()) {
      _startAutoplay();
    }
  }

  void _startAutoplay() {
    assert(_timer == null, "Timer must be stopped before start!");
    _timer =
        Timer.periodic(Duration(milliseconds: widget.autoplayDelay), _onTimer);
  }

  void _onTimer(Timer timer) {
    _controller.next(animation: true);
  }

  void _stopAutoplay() {
    if (_timer != null) {
      _timer!.cancel();
      _timer = null;
    }
  }
}

class _SliderCarouselState extends _SliderCarouselTimerMixin {
  late int _activeIndex;

  TransformerPageController? _pageController;
  @override
  void initState() {
    _activeIndex = widget.index ?? 0;
    super.initState();
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
  }

  @override
  void didUpdateWidget(SliderCarousel oldWidget) {
    super.didUpdateWidget(oldWidget);

    scheduleMicrotask(() {
      // So that we have a chance to do `removeListener` in child widgets.
      if (_pageController != null) {
        _pageController?.dispose();
        _pageController = null;
      }
    });
    if (widget.index != null && widget.index != _activeIndex) {
      _activeIndex = widget.index!;
    }
  }

  void _onIndexChanged(int index) {
    if (mounted) {
      setState(() {
        _activeIndex = index;
      });
    }
    if (widget.onIndexChanged != null) {
      widget.onIndexChanged!(index);
    }
  }

  Widget _buildSliderCarousel() {
    late IndexedWidgetBuilder itemBuilder;
    itemBuilder = widget.itemBuilder!;

    return new _StackSliderCarousel(
      loop: widget.loop,
      itemWidth: widget.itemWidth!,
      itemHeight: widget.itemHeight!,
      itemCount: widget.itemCount,
      itemBuilder: itemBuilder,
      index: _activeIndex,
      curve: widget.curve,
      duration: widget.duration,
      onIndexChanged: _onIndexChanged,
      controller: _controller,
      scrollDirection: widget.scrollDirection,
      offsetCenter: widget.offsetCenter,
      scaleCenter: widget.scaleCenter,
    );
  }

  @override
  Widget build(BuildContext context) {
    Widget sliderCarousel = _buildSliderCarousel();
    List<Widget>? listForStack;
    if (listForStack != null) {
      return new Stack(
        children: listForStack,
      );
    }

    return sliderCarousel;
  }
}

abstract class _SubSliderCarousel extends StatefulWidget {
  final IndexedWidgetBuilder? itemBuilder;
  final int? itemCount;
  final int? index;
  final ValueChanged<int>? onIndexChanged;
  final SliderCarouselController? controller;
  final int? duration;
  final Curve? curve;
  final double? itemWidth;
  final double? itemHeight;
  final double offsetCenter;
  final double? scaleCenter;
  final bool? loop;
  final Axis? scrollDirection;

  _SubSliderCarousel(
      {Key? key,
      this.loop,
      this.itemHeight,
      this.itemWidth,
      this.duration,
      this.curve,
      this.itemBuilder,
      this.controller,
      this.index,
      this.itemCount,
      this.scrollDirection: Axis.horizontal,
      this.offsetCenter: 0,
      this.scaleCenter,
      this.onIndexChanged})
      : super(key: key);

  @override
  State<StatefulWidget> createState();

  int getCorrectIndex(int indexNeedsFix) {
    if (itemCount == 0) return 0;
    int value = indexNeedsFix % itemCount!;
    if (value < 0) {
      value += itemCount!;
    }
    return value;
  }
}

class _StackSliderCarousel extends _SubSliderCarousel {
  _StackSliderCarousel({
    Key? key,
    Curve? curve,
    int? duration,
    SliderCarouselController? controller,
    ValueChanged<int>? onIndexChanged,
    double? itemHeight,
    double? itemWidth,
    IndexedWidgetBuilder? itemBuilder,
    int? index,
    bool? loop,
    int? itemCount,
    Axis? scrollDirection,
    required double offsetCenter,
    double? scaleCenter,
  }) : super(
            loop: loop,
            key: key,
            itemWidth: itemWidth,
            itemHeight: itemHeight,
            itemBuilder: itemBuilder,
            curve: curve,
            duration: duration,
            controller: controller,
            index: index,
            onIndexChanged: onIndexChanged,
            itemCount: itemCount,
            offsetCenter: offsetCenter,
            scaleCenter: scaleCenter,
            scrollDirection: scrollDirection);

  @override
  State<StatefulWidget> createState() {
    return new _StackViewState();
  }
}

class _StackViewState extends _CustomLayoutStateBase<_StackSliderCarousel> {
  late List<double> scales = [0.0, 1.0, 1.0, widget.scaleCenter ?? 0, 0.0];
  late List<double> offsets = [
    -(_sliderWidth - widget.itemWidth!) / 2,
    (_sliderWidth - widget.itemWidth!) / 2 * widget.offsetCenter,
    -(_sliderWidth - widget.itemWidth!) / 2 * widget.offsetCenter,
    -scales[3] > 1 ? (widget.itemWidth! * (scales[3] - 1)) / 2 : 0,
    _sliderHeight
  ];
  late List<double> opacity = [0.0, 1.0, 1.0, 1.0, 0.0];

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
  }

  void _updateValues() {
    if (widget.scrollDirection == Axis.horizontal) {
      double space = (_sliderWidth - widget.itemWidth!) / 2;
      double spaceItem =
          scales[3] > 1 ? (widget.itemWidth! * (scales[3] - 1)) / 2 : 0;
      offsets = [
        -space,
        space * widget.offsetCenter,
        -space * widget.offsetCenter,
        -spaceItem,
        _sliderWidth
      ];
    } else {
      double space = (_sliderHeight - widget.itemHeight!) / 2;
      double spaceItem =
          scales[3] > 1 ? (widget.itemHeight! * (scales[3] - 1)) / 2 : 0;
      offsets = [
        -space,
        space * widget.offsetCenter,
        -space * widget.offsetCenter,
        -spaceItem,
        _sliderHeight
      ];
    }
  }

  @override
  void didUpdateWidget(_StackSliderCarousel oldWidget) {
    _updateValues();
    super.didUpdateWidget(oldWidget);
  }

  @override
  void afterRender() {
    super.afterRender();
    //length of the values array below
    _animationCount = 5;
    //Array below this line, '0' index is 1.0 ,witch is the first item show in SliderCarousel.
    _startIndex = -3;
    scales = [0.0, 1.0, 1.0, widget.scaleCenter ?? 0, 0.0];
    opacity = [0.0, 1.0, 1.0, 1.0, 0.0];

    _updateValues();
  }

  @override
  Widget _buildItem(int i, int realIndex, double animationValue) {
    double s = _getValue(scales, animationValue, i);
    double f = _getValue(offsets, animationValue, i);
    double o = _getValue(opacity, animationValue, i);

    Offset offset = widget.scrollDirection == Axis.horizontal
        ? new Offset(f, 0.0)
        : new Offset(0.0, f);

    Alignment alignment = widget.scrollDirection == Axis.horizontal
        ? Alignment.centerLeft
        : Alignment.topCenter;

    return new Opacity(
      opacity: o,
      child: new Transform.translate(
        key: new ValueKey<int>(_currentIndex + i),
        offset: offset,
        child: new Transform.scale(
          scale: s,
          alignment: alignment,
          child: new SizedBox(
            width: widget.itemWidth ?? double.infinity,
            height: widget.itemHeight ?? double.infinity,
            child: widget.itemBuilder!(context, realIndex),
          ),
        ),
      ),
    );
  }
}


lib/src/slider_controller.dart
import 'package:mafya/core/init/transformer_page_view/another_transformer_page_view.dart';

class SliderCarouselController extends IndexController {
  // Autoplay is started
  static const int START_AUTOPLAY = 2;

  // Autoplay is stopped.
  static const int STOP_AUTOPLAY = 3;

  // Indicate that the user is swiping
  static const int SWIPE = 4;

  // Indicate that the `SliderCarousel` has changed it's index and is building it's ui ,so that the
  // `SliderCarouselPluginConfig` is available.
  static const int BUILD = 5;

  // available when `event` == SliderCarouselController.SWIPE
  // this value is PageViewController.pos
  late double pos;

  int? index;
  late bool animation;
  bool? autoplay;

  SliderCarouselController();

  void startAutoplay() {
    event = SliderCarouselController.START_AUTOPLAY;
    this.autoplay = true;
    notifyListeners();
  }

  void stopAutoplay() {
    event = SliderCarouselController.STOP_AUTOPLAY;
    this.autoplay = false;
    notifyListeners();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter_slidercarousel/flutter_slidercarousel.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo SliderCarousel',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Flutter SliderCarousel'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

List<Color> colors = [
  Colors.blue,
  Colors.red,
  Colors.yellow,
  Colors.green,
  Colors.pink
];

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: SliderCarousel(
          itemCount: 5,
          onIndexChanged: (index)=>print("onIndexChanged: "+index.toString()),
          onTap: (index)=>print("onTap: "+index.toString()),
          itemBuilder: (context, idx) {
            return Container(
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                color: colors[idx],
              ),
            );
          },
          offsetCenter: 0.75,
          scaleCenter: 1.25,
          itemWidth: 150,
          itemHeight: 150,
        ),
      ),
    );
  }
}
